pmlc/rt/level_zero/level_zero_invocation.cc

```diff
@@ -0,0 +1,191 @@
+// Copyright 2020 Intel Corporation
+#include "pmlc/rt/level_zero/level_zero_invocation.h"
+
+#include <algorithm>
+#include <chrono>
+#include <string>
+#include <utility>
+
+#include "pmlc/util/logging.h"
+
+namespace pmlc::rt::level_zero {
+
+void LevelZeroMemory::enqueueRead(ze_command_list_handle_t list, void *dst,
+                                  std::vector<ze_event_handle_t> &dependencies,
+                                  ze_event_handle_t &resultE) {
+  lzu::append_memory_copy(list, dst, buffer, bytes, resultE,
+                          dependencies.size(), dependencies.data());
+}
+
+void LevelZeroMemory::enqueueWrite(ze_command_list_handle_t list, void *src,
+                                   std::vector<ze_event_handle_t> &dependencies,
+                                   ze_event_handle_t &resultE) {
+  lzu::append_memory_copy(list, buffer, src, bytes, resultE,
+                          dependencies.size(), dependencies.data());
+}
+
+LevelZeroKernel::LevelZeroKernel(ze_module_handle_t module, std::string name)
+    : module(module), name(name) {
+  kernel = lzu::create_function(module, /*flag*/ 0, name);
+}
+
+LevelZeroKernel::~LevelZeroKernel() {
+  lzu::destroy_function(kernel);
+  lzu::destroy_module(module);
+}
+
+void LevelZeroKernel::addDependency(LevelZeroEvent *event) {
+  dependencies.push_back(event->getEvent());
+}
+
+void LevelZeroKernel::setArg(unsigned idx, LevelZeroMemory *memory) {
+  void *buf = reinterpret_cast<int64_t *>(memory->getBuffer());
+  lzu::set_argument_value(kernel, idx, sizeof(buf), &buf);
+}
+
+void LevelZeroKernel::enqueue(ze_command_list_handle_t list,
+                              ze_group_count_t gws, ze_group_count_t lws,
+                              ze_event_handle_t &resultE) {
+  // The usage of gws and lws will depend on their meaning.
```

Comment:
I think this comments is good for us. But for PR to plaidml, we can remove this comment.

Reply:
Or you can replace this comments by the meaning of gws in lzu::append_launch_function

Reply:
maybe add a new line for unused variable like `(void)lws;` or define a macro then `UNUSED(lws)`;

Reply:
Add UNUSED_VARIABLE with (void)(x),  may be save space.     Reword and move comment to enqueueKennel function.